crux-dbg/src/Crux/Debug/LLVM.hs

```diff
@@ -150,9 +172,78 @@ do_breakpoint :: (wptr ~ CLE.ArchWidth arch)
 
+-- TODO: export in crux
+lookupString :: (LCB.IsSymInterface sym, CLM.HasLLVMAnn sym, CLO.ArchOk arch)
+             => LCS.GlobalVar CLM.Mem
+             -> LCS.RegEntry sym (CLT.LLVMPointerType (CLE.ArchWidth arch))
+             -> C.OverM personality sym (CLI.LLVM arch) String
+lookupString mvar ptr =
+  do sym <- LCS.getSymInterface
+     mem <- LCS.readGlobal mvar
+     bytes <- liftIO (CLM.loadString sym mem (regValue ptr) Nothing)
+     return (BS8.unpack (BS.pack bytes))
+
+-- TODO: export in crux
+withSolverAdapter :: CCS.SolverOffline -> (WS.SolverAdapter st -> a) -> a
+withSolverAdapter solverOff k =
+  case solverOff of
+    CCS.Boolector -> k WS.boolectorAdapter
+    CCS.DReal -> k WS.drealAdapter
+    CCS.SolverOnline CCS.CVC4 -> k WS.cvc4Adapter
+    CCS.SolverOnline CCS.STP -> k WS.stpAdapter
+    CCS.SolverOnline CCS.Yices -> k WS.yicesAdapter
+    CCS.SolverOnline CCS.Z3 -> k WS.z3Adapter
+
+parseSolverOffline :: C.CruxOptions -> CCS.SolverOffline
+parseSolverOffline cruxOpts =
+  case CCS.parseSolverConfig cruxOpts of
+    Right (CCS.SingleOnlineSolver _onSolver) -> CCS.SolverOnline _onSolver
+    Right (CCS.OnlineSolverWithOfflineGoals _ _offSolver) -> _offSolver
+    Right (CCS.OnlyOfflineSolver _offSolver) -> _offSolver
+    Right (CCS.OnlineSolverWithSeparateOnlineGoals _ _onSolver) -> CCS.SolverOnline _onSolver
+    Left _ -> CCS.SolverOnline CCS.Yices
+
+do_debug_assert :: ( CLO.ArchOk arch
+                  , LCB.IsSymInterface sym
+                  , CLM.HasLLVMAnn sym
+                  , sym ~ (WEB.ExprBuilder t st fs)
+                  , SC.Architecture arch' t
+                  , SC.SymbolicArchitecture arch' t
+                  , ext ~ CLI.LLVM arch
+                  )
+                => CCS.SolverOffline
+                -> SB.DebuggerConfig t ext arch'
+                -> LCS.GlobalVar CLM.Mem
+                -> sym
+                -> Ctx.Assignment (LCS.RegEntry sym) (Ctx.EmptyCtx Ctx.::> LCT.BVType 8 Ctx.::> CLT.LLVMPointerType (CLE.ArchWidth arch) Ctx.::> LCT.BVType 32)
+                -> C.OverM personality sym (CLI.LLVM arch) (LCS.RegValue sym LCT.UnitType)
+do_debug_assert offSolver sconf mvar sym (Ctx.Empty Ctx.:> p Ctx.:> pFile Ctx.:> line) = do
+  cond <- liftIO $ WI.bvIsNonzero sym (regValue p)
+  file <- lookupString mvar pFile
+  l <- case WI.asBV (regValue line) of
+         Just (BV.BV l)  -> return (fromInteger l)
+         Nothing -> return 0
+  let pos = WPL.SourcePos (T.pack file) l 0
+  loc <- liftIO $ WI.getCurrentProgramLoc sym
+  let loc' = loc{ WPL.plSourceLoc = pos }
+  let msg = LCS.GenericSimError "crucible_debug_assert"
+  ret <- liftIO $ LCB.addDurableAssertion sym (LCB.LabeledPred cond (LCS.SimError loc' msg))
+  simState <- CMS.get
+  liftIO . withSolverAdapter offSolver $ \adapter -> do
+    let logData = WS.defaultLogData
+    negCond <- WI.notPred sym cond
+    pathCond <- LCB.getPathCondition sym
+    testCond <- WI.impliesPred sym pathCond negCond
```

Comment:
What is the rationale behind `impliesPred` instead of just `andPred` here?

Reply:
I'm testing for the satisfiability of `path ⇒ ¬cond` which is boolean equivalent to `¬(path ∧ cond)`, I just thought the implication is more readable. A model for this predicate would refute either `path` or `cond`.

Reply:
I don't think we ever need to refute `path` - it just contains assumptions, so we can't really refute them.  Also see how crux handles it (`path ∧ ¬cond`): https://github.com/GaloisInc/crucible/blob/master/crux/src/Crux/Goal.hs#L236

Reply:
Got it, updated.